File: financial-domain-starter/src/main/java/com/nixsolutions/financial/discovery/ServiceRegistryAware.java
package com.nixsolutions.financial.discovery;

import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponentsBuilder;

public interface ServiceRegistryAware
{
  default WebClient getWebClient(String uri)
  {
    return WebClient.builder()
        .baseUrl(uri)
        .build();
  }

  default String composePath(String... pathElements)
  {
    UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromHttpUrl(getServiceUrl());
    uriComponentsBuilder.pathSegment(pathElements);
    return uriComponentsBuilder.toUriString();
  }

  String getServiceUrl();
}

File: financial-domain-starter/src/main/kotlin/com/nixsolutions/financial/discovery/ServiceRegistryAware.kt
package com.nixsolutions.financial.discovery

import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.util.UriComponentsBuilder

interface ServiceRegistryAware {

  fun getServiceUrl(): String;

  @JvmDefault
  fun getWebClient(uri: String): WebClient {
    return WebClient.builder()
        .baseUrl(uri)
        .build()
  }

  @JvmDefault
  fun composePath(vararg pathElements: String): String {
    val uriComponentsBuilder = UriComponentsBuilder.fromHttpUrl(getServiceUrl())
    uriComponentsBuilder.pathSegment(*pathElements)
    return uriComponentsBuilder.toUriString()
  }
}